src/jwe/envelope.rs

```diff
@@ -0,0 +1,174 @@
+use serde::{Deserialize, Serialize};
+use serde_enum_str::{Deserialize_enum_str, Serialize_enum_str};
+use serde_json::Value;
+
+/// Subset of JWE in generic json serialization form used for authcrypt
+/// and anoncrypt message types.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub(crate) struct JWE<'a> {
+    /// BASE64URL(UTF8(JWE Protected Header))
+    /// Note: this field value is used as AAD for JWE Ciphertext
+    pub protected: &'a str,
+
+    /// Array of recipient-specific objects
+    pub recipients: Vec<Recepient<'a>>,
+
+    /// BASE64URL(JWE Initialization Vector)
+    pub iv: &'a str,
+
+    /// BASE64URL(JWE Ciphertext)
+    pub ciphertext: &'a str,
+
+    /// BASE64URL(JWE Authentication Tag)
+    pub tag: &'a str,
+}
+
+/// Protected header for authcrypt/anoncrypt-specific JWE.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub(crate) struct ProtectedHeader<'a> {
+    /// Must be `application/didcomm-encrypted+json` or `didcomm-encrypted+json` for now.
+    /// Something like `application/didcomm-encrypted+cbor` can be introduced in the
+    /// future.
+    pub typ: Option<&'a str>,
+
+    /// Cryptographic algorithm used to encrypt or determine the value of the CEK.
+    pub alg: Algorithm,
+
+    /// Identifies the content encryption algorithm used to perform authenticated encryption
+    /// on the plaintext to produce the ciphertext and the Authentication Tag.
+    pub enc: EncAlgorithm,
+
+    /// Sender KID as DID Url.
+    /// If absent implementations MUST be able to resolve the sender kid from the `apu` header.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub skid: Option<&'a str>,
+
+    /// BASE64URL("skid" header value),
```

Comment:
Shall we add #[serde(skip_serializing_if = "Option::is_none")] here?

Reply:
Yes, fixed. Thanks for noticing this.